psh/dd/dd.c

```diff
@@ -0,0 +1,342 @@
+		"\tbs=BYTE   read/write block size of BYTES bytes\n"
+		"\tcount=N   copy only N input blocks\n"
+		"\tseek=N    skip N bs-sized blocks at start of output\n"
+		"\tskip=N    skip N bs-sized blocks at start of input\n");
+}
+
+
+static ssize_t getnumber(char *numstr)
+{
+	ssize_t value;
+
+	if (isdigit(*numstr) == 0) {
+		return -EINVAL;
+	}
+
+	value = 0;
+	while (isdigit(*numstr) != 0) {
+		value = value * 10 + *(numstr++) - '0';
+	}
+
+	switch (*(numstr++)) {
+		case 'M':
+			value *= 1024 * 1024;
+			break;
+
+		case 'k':
+			value *= 1024;
+			break;
+
+		case 'b':
+			value *= 512;
+			break;
+
+		case 'w':
+			value *= 2;
+			break;
+
+		case 'c':
+			/* value *= 1; */
+			break;
+
+		case '\0':
+			return value;
+
+		default:
+			return -1;
+	}
+
+	if (*numstr != '\0') {
+		return -EINVAL;
+	}
+
+	return value;
+}
+
+
+static int psh_dd(int argc, char **argv)
+{
+	const struct param_s *par;
+	char *buf, *str, *cp;
+	int infd, outfd, incc, outcc;
+	ssize_t intotal, outtotal;
+
+	char *infile = NULL;
+	char *outfile = NULL;
+	ssize_t count = SSIZE_MAX;
+	ssize_t seekval = 0;
+	ssize_t skipval = 0;
+	ssize_t inmax = 0;
+	int err = 0, blocksz = 512;
+
+	if ((argc > 1) && (argv[1][0] == '-') && (argv[1][1] == 'h')) {
+		psh_dd_usage();
+		return EXIT_SUCCESS;
+	}
+
+	while (--argc > 0) {
+		str = *(++argv);
+		cp = strchr(str, '=');
+		if (cp == NULL) {
+			fprintf(stderr, "Bad dd argument\n");
+			return EXIT_FAILURE;
+		}
+
+		for (par = params; par->name; par++) {
+			if (strncmp(str, par->name, cp - str) == 0) {
+				break;
+			}
+		}
+
+		cp++;
+
+		switch (par->value) {
+			case param_if:
+				if (infile != NULL) {
+					fprintf(stderr, "Multiple input files illegal\n");
+					return EXIT_FAILURE;
+				}
+				infile = cp;
+				break;
+
+			case param_of:
+				if (outfile != NULL) {
+					fprintf(stderr, "Multiple output files illegal\n");
+					return EXIT_FAILURE;
+				}
+				outfile = cp;
+				break;
+
+			case param_bs:
+				blocksz = getnumber(cp);
+				if (blocksz <= 0) {
+					fprintf(stderr, "Bad block size value\n");
+					return EXIT_FAILURE;
+				}
+				break;
+
+			case param_count:
+				count = getnumber(cp);
+				if (count < 0) {
+					fprintf(stderr, "Bad count value\n");
+					return EXIT_FAILURE;
+				}
+				break;
+
+			case param_seek:
+				seekval = getnumber(cp);
+				if (seekval < 0) {
+					fprintf(stderr, "Bad seek value\n");
+					return EXIT_FAILURE;
+				}
+				break;
+
+			case param_skip:
+				skipval = getnumber(cp);
+				if (skipval < 0) {
+					fprintf(stderr, "Bad skip value\n");
+					return EXIT_FAILURE;
+				}
+				break;
+
+			default:
+				fprintf(stderr, "Unknown dd parameter\n");
+				return EXIT_FAILURE;
+		}
+	}
+
+	buf = malloc(blocksz);
+	if (buf == NULL) {
+		fprintf(stderr, "Cannot allocate buffer\n");
+		return EXIT_FAILURE;
+	}
+
+	intotal = 0;
+	outtotal = 0;
+
+	infd = ((infile == NULL) ? fileno(stdin) : open(infile, O_RDONLY));
+	if (infd < 0) {
+		perror(infile);
```

Comment:
More readable message, especially if empty file name `if=`.
```suggestion
		fprintf(stderr, "'%s': %s\n", infile, strerror(errno));
```

Reply:
I like it. Thanks.